Print ebinds in FQ and fix crash with nested quantifiers

atondwal opened the pull request.

atondwal committed:
Print out ebinds

atondwal committed:
Add todo/ebind-elim-crash.fq

This is a crash that occurs when using both ebinds and --eliminate=some

atondwal committed:
Added NFData instance for Sol

It's really quite helpful when it comes to debugging

atondwal committed:
Fix crash with ebinds

atondwal committed:
Added another test case for elim-ebinds crash.

atondwal committed:
update tests

atondwal committed:
separate TODO into crash and incompleteness

atondwal committed:
fix circle.yml

ranjitjhala commented:
Please don’t merge into develop until and unless the LH test also work.

On Mon, Apr 30, 2018 at 1:59 PM Anish Tondwalkar <notifications@github.com>
wrote:

> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/ucsd-progsys/liquid-fixpoint/pull/363
> Commit Summary
>
>    - Print out ebinds
>    - Add todo/ebind-elim-crash.fq
>    - Added NFData instance for Sol
>    - Fix crash with ebinds
>    - Added another test case for elim-ebinds crash.
>    - update tests
>    - separate TODO into crash and incompleteness
>
> File Changes
>
>    - *M* src/Language/Fixpoint/Solver/Solution.hs
>    <https://github.com/ucsd-progsys/liquid-fixpoint/pull/363/files#diff-0>
>    (8)
>    - *M* src/Language/Fixpoint/Solver/Solve.hs
>    <https://github.com/ucsd-progsys/liquid-fixpoint/pull/363/files#diff-1>
>    (57)
>    - *M* src/Language/Fixpoint/Types/Constraints.hs
>    <https://github.com/ucsd-progsys/liquid-fixpoint/pull/363/files#diff-2>
>    (9)
>    - *M* src/Language/Fixpoint/Types/Environments.hs
>    <https://github.com/ucsd-progsys/liquid-fixpoint/pull/363/files#diff-3>
>    (14)
>    - *M* src/Language/Fixpoint/Types/Solutions.hs
>    <https://github.com/ucsd-progsys/liquid-fixpoint/pull/363/files#diff-4>
>    (21)
>    - *A* tests/neg/ebind-elim2.fq
>    <https://github.com/ucsd-progsys/liquid-fixpoint/pull/363/files#diff-5>
>    (26)
>    - *M* tests/test.hs
>    <https://github.com/ucsd-progsys/liquid-fixpoint/pull/363/files#diff-6>
>    (3)
>    - *R* tests/todo-crash/wl01.fq
>    <https://github.com/ucsd-progsys/liquid-fixpoint/pull/363/files#diff-7>
>    (0)
>    - *A* tests/todo/ebind-elim-crash.fq
>    <https://github.com/ucsd-progsys/liquid-fixpoint/pull/363/files#diff-8>
>    (41)
>
> Patch Links:
>
>    - https://github.com/ucsd-progsys/liquid-fixpoint/pull/363.patch
>    - https://github.com/ucsd-progsys/liquid-fixpoint/pull/363.diff
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/ucsd-progsys/liquid-fixpoint/pull/363>, or mute the
> thread
> <https://github.com/notifications/unsubscribe-auth/ABkuOBQ_bTxO9pyyv1ooqyjgjIJ08Ctfks5tt3JtgaJpZM4TtKap>
> .
>

atondwal commented:
It fixes the LH crashes (make one of the tests crash -> SAFE and another from crash -> UNSAFE), but I was going to hold off until you reviewed the LH PR and merge them together anyways

ranjitjhala commented:
:thumbsup:

On Mon, Apr 30, 2018 at 2:17 PM, Anish Tondwalkar <notifications@github.com>
wrote:

> It fixes the LH crashes (make one of the tests crash -> SAFE and another
> from crash -> UNSAFE), but I was going to hold off until you reviewed the
> LH PR and merge them together anyways
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/ucsd-progsys/liquid-fixpoint/pull/363#issuecomment-385531749>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABkuONqTSzwTYEtDpGTM_m8QhU7_nnMnks5tt39vgaJpZM4TtKap>
> .
>

atondwal committed:
Minimzed failing test case futher

ebind-elim-crash is dervied from Implicit2.hs.

It had two bugs in it:

 - a CRASH resulting from twice-bound ebinds with eliminate, fixed
   by dropping the inner quantifer (see tests/neg/ebind-elim2.fq)

 - The current bug which causes eliminate to incorrectly mark the
   file as UNSAFE

ranjitjhala commented on src/Language/Fixpoint/Solver/Solve.hs:
```diff
@@ -301,25 +301,54 @@ partitionInfo (i, fi)
-exElim env xi p = F.notracepp msg (F.pExist yts p) 
-  where 
+exElim :: F.SEnv (F.BindId, F.Sort) -> F.BindId -> F.Symbol -> F.Pred -> F.Pred
+exElim env xi x p = F.notracepp msg (F.pExist yts (clearExist x p))
```
What is this `clearExist` for?

ranjitjhala commented:
Ok, am going to merge and then UNDO the `clearExists`.
